### PR TITLE
feat(files): add keyboard shortcuts

### LINKED ISSUE #1444: Keyboard shortcut support
A few apps have awesome keyboard shortcuts, which makes it super easy and fast to use nextcloud. I think we should define standards for nextcloud, so the shortcuts are consistent.

*Additionally It would be nice to have a current shortcut overview help like github does it.
(Try it by pushing the `?` key)
Could this work as an app?*

---
### Advised general shortcuts
| Shortcut | Action |
| --- | --- |
| <kbd>r</kbd> | Reload messages / feed |
| <kbd>▶︎</kbd> / <kbd>j</kbd> /  <kbd>n</kbd> | Next item / file |
| <kbd>◀︎</kbd> / <kbd>k</kbd> /  <kbd>p</kbd> | Previous message / item / file |
| <kbd>del</kbd> | Delete message |
| <kbd>o</kbd> / <kbd>enter</kbd> | open Item |
| <kbd>q</kbd> | Focus internal search |
| <kbd>u</kbd> | Toggle read/unread |
| <kbd>s</kbd> | Toggle Star (flag) message |
| <kbd>c</kbd> | create item |
| <kbd>esc</kbd> | close menu |

**Items** can be mails, files, folders, etc., if not stated otherwise

### Shortcuts/combinations that should be avoided
- Combinations with <kbd>ctrl</kbd> and <kbd>cmd</kbd> in general
- <kbd>▼</kbd> and <kbd>▲</kbd>, because of interference with scrolling

### Undecided shortcuts from News App
Maybe to special to generalize them

Shortcut | Action
--- | ---
<kbd>e</kbd> | expandItem
<kbd>i</kbd> /  <kbd>l</kbd> | toggleStar
<kbd>h</kbd> | toggleStar && scrollToNextItem
<kbd>f</kbd> | nextFeed
<kbd>d</kbd> | previousFeed
<kbd>c</kbd> | previousFolder
<kbd>a</kbd> | scrollToActiveNavigationEntry
<kbd>v</kbd> | nextFolder
<kbd>shift</kbd> + <kbd>a</kbd> | markAllRead

### Files App

| Shortcut | Action |
| --- | --- |
| <kbd>ctrl</kbd>/<kbd>cmd</kbd> + <kbd>n</kbd> | new folder |
| <kbd>ctrl</kbd>/<kbd>cmd</kbd> + <kbd>shift</kbd> + <kbd>n</kbd> | new file |
| <kbd>esc</kbd> | close menu (while new file context menu is open) |
| <kbd>up</kbd>/<kbd>down</kbd> | select file/folder |
| <kbd>enter</kbd> | open file/folder |
| <kbd>delete</kbd>/<kbd>backspace</kbd> | delete file/folder |
| <kbd>ctrl</kbd>/<kbd>cmd</kbd> + <kbd>f</kbd> | focus search field |
| <kbd>esc</kbd> | close menu |

### Mail App

| Shortcut | Action |
| --- | --- |
| <kbd>r</kbd> | Reload messages |
| <kbd>◀︎</kbd> | Previous message |
| <kbd>▶︎</kbd> | Next message |
| <kbd>s</kbd> | Star (flag) message |
| <kbd>u</kbd> | Toggle read/unread |
| <kbd>del</kbd> | Delete message |

### News App

Shortcut | Action
---|---
<kbd>j</kbd> /  <kbd>n</kbd> /  <kbd>▶︎</kbd> | scrollToNextItem
<kbd>k</kbd> /  <kbd>p</kbd> /  <kbd>◀︎</kbd> | scrollToPreviousItem
<kbd>u</kbd> | toggleUnread
<kbd>e</kbd> | expandItem
<kbd>s</kbd> /  <kbd>i</kbd> /  <kbd>l</kbd> | toggleStar
<kbd>h</kbd> | toggleStar && scrollToNextItem
<kbd>o</kbd> | openLink
<kbd>r</kbd> | reloadFeed
<kbd>f</kbd> | nextFeed
<kbd>d</kbd> | previousFeed
<kbd>c</kbd> | previousFolder
<kbd>a</kbd> | scrollToActiveNavigationEntry
<kbd>v</kbd> | nextFolder
<kbd>q</kbd> | Focus search
<kbd>shift</kbd> + <kbd>a</kbd> | markAllRead

## Todo:
- [x] collect the shortcuts that are already available
- [ ] general: improve tabindex
- [ ] general: shortcuts for navigation / menu 
  - [ ]  jump to apps menu
  - [ ]  jump to sidebar
  - [ ]  jump to settings menu
  - [ ]  select menu point (sidebar/apps/settings) with arrow keys
- [ ] add a dedicated page in the documentation
- [ ] files app: add file option (maybe distinct issue)
  - [ ]  rename
  - [ ]  open and focus details
  - [ ]  share (is an essential feature and should be able to be  directly executed on a selected file)
  - [ ] descend into newly created folder by clicking enter again [#1905]
- [ ] contacts app
- [ ] calendar app [nextcloud/calendar#157]
- [ ] gallery app
- [ ] ?app for a helper listing shortcuts in the current view

### LINKED ISSUE #30484: Keyboard shortcuts for common actions in Files
Files right now is not super accessible by keyboard, with other software like Gdrive, Onedrive and Dropbox having nice keyboard shortcuts for common actions. It would be great if NC also had better keyboard accessibility. Here are the most common actions and their keyboard shortcuts across different software, and any similar actions in other Nextcloud apps also. Since there are sometimes slightly different shortcuts for different actions across softwares, we can't have something that's 100% familiar to everyone, but we can use the most common shortcut key.

| Action                  | Gdrive                                            | Dropbox      | Onedrive    | Other Nextcloud apps | Recommended shortcut                            |
|-------------------------|---------------------------------------------------|--------------|-------------|----------------------|-------------------------------------------------|
| Navigate through files  | ↑ ↓ → ← / k j h l                                 | ↑ ↓          | ↑ ↓ → ←     | ↑ ↓ in Mail          | ↑ ↓ → ← (left and right to support grid view)   |
| Select/unselect         | Ctrl+Space or x                                   |              | Space       |                      | Ctrl+Space                                      |
| Select range            | Shift+↑ ↓ → ←                                     |              |             |                      |                                                 |
| Select all              | Ctrl + a                                          | Ctrl + a     | Ctrl + a    |                      | Ctrl + a                                        |
| Unselect all            | Ctrl + n                                          | Esc          | Ctrl + d    |                      | Esc (Ctrl + n is used in Firefox for opening a new window, so Esc can be used for unselect as long as no popup is open)                                 |
| Search                  | /                                                 | /            | /           | / for quick find     | /                                               |
| Delete                  | Del                                               | Del          | Del         |                      | Del                                             |
| Rename                  | n                                                 | F2           | F2          |                      | F2                                              |
| Open file/folder        | Enter or o                                        | →            | Enter       | Enter in Mail        | Enter                                           |
| Close open file/popup   | Esc                                               |              | Esc         |                      | Esc                                             |
| Go up a folder          | g p                                               | ←            | Alt + ↑     |                      | Alt + ↑                                         |
| Upload                  | Shift + u (upload file), Shift + i (upload folder) |              |             |                      | Shift + u (open upload menu)                    |
| Share                   | .                                                 |              | s           |                      | . (s is used for favouriting)                                               |
| Favourite/star  | s                                               |              |            | s in Mail                     | s                                               |
| Toggle sidebar/details  | g d                                               |              | i           |                      | d                                               |
| Open actions menu       | a                                                 |              | Shift + F10 |                      | a (we could implement this across all apps)                               |
| Cut/copy/paste          | Ctrl + x/c/v                                      | Ctrl + x/c/v |             |                      | Ctrl + x/c/v                                    |
| Switch grid/list view   | v                                                 |              |             |                      | v                                               |
| Go to navigation        | g n                                               |              |             |                      | n (we could implement this across all the apps) |
| Show keyboard shortcuts | ?                                                 | ?            | ?           |                      | ?                                               |

What do you think? 
cc @jancborchardt who also worked on this and @skjnldsv for Files

### PR DESCRIPTION
**Pull request by skjnldsv**
Fix #1444
Fix #30484

| 1 | 2 |
|:--:|:--:|
|||

<details>
<summary>Shortcuts</summary>
### Selection
- [x] Select all
- [x] Unselect all
- [x] Select/unselect current file

### Actions
- [x] Open actions menu
- [x] Delete
- [x] Favourite/star
- [x] Rename
- [x] Toggle sidebar/details

### Navigation
- [x] Go up a folder
- [x] Navigate through files
- [x] Open file/folder

### Usability
- [ ] Open upload menu
- [x] Show keyboard shortcuts
- [x] Switch grid/list view

### Later
- [ ] Search
- [ ] Share (suggested shortcut is ... a dot ? :thinking: https://github.com/nextcloud/server/issues/30484#issuecomment-2547897036 )
- [ ] Go to navigation (should be done in https://github.com/nextcloud/nextcloud-vue)
- [ ] Cut/copy/paste https://github.com/nextcloud/server/issues/49892
</details>

## Status
- [x] Actions
- [x] Tests
- [ ] Investigate performance loss while scrolling to file ID

### CODE REVIEW

**Comment by skjnldsv** [comment]
Alright, first PR is ready.
Some more will follow for tests and some tiny changes

**Review by ShGKme — DISMISSED** [review]
Cool! I really like the shortcuts!

But there are some issues:
1. This is not clear for me when a file is in focus and when it is not. For example, for arrows and CTRL+Space hotkeys. Especially because the click on the file executes an action rather than selecting it.
2. We should check for disabled shortcuts for a11y (many keys are used in assistive software). If I remember correctly, this is why `OCP.Accessibility.disableKeyboardShortcuts` was added in the past
3. It doesn't check for the current element. As a result, it is not possible to make any input anymore, for example, in text fields.
    
https://github.com/user-attachments/assets/f0fe73f9-c134-4e11-907a-b7ac7c0dd572

Btw, we have [useHotKey](https://nextcloud-vue-components.netlify.app/#/Composables/useHotKey) composable to reuse global shortcuts implementation. It might fix some of the issues and reduce copy-paste like `mounted + beforeDestroy` hooks.

**Comment by skjnldsv** [comment]
> Btw, we have [useHotKey](https://nextcloud-vue-components.netlify.app/#/Composables/useHotKey) composable to reuse global shortcuts implementation. It might fix some of the issues and reduce copy-paste like `mounted + beforeDestroy` hooks.

TIL, thanks :)
Yeah, I just realized the input too!
Btw, we should probably check for `role=dialig` like I did, instead of `.modal-mask`
https://github.com/nextcloud/nextcloud-vue/blob/10626766712de1e5858b0f71a9babd552bfa6805/src/composables/useHotKey/index.js#L26

File: apps/files/src/actions/deleteAction.ts [hunk 1]
```diff
@@ -41,8 +43,14 @@ export const action = new FileAction({
 		try {
 			let confirm = true
 
+			// Trick to detect if the action was called from a keyboard event
+			// we need to make sure the method calling have its named containing 'keydown'
+			// here we use `onKeydown` method from the FileEntryActions component
+			const callStack = new Error().stack || ''
+			const isCalledFromEventListener = callStack.toLocaleLowerCase().includes('keydown')
+
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply apps/files/src/actions/deleteAction.ts:51]: It IS very very tricky! :magic_wand: 
The exec is a direct API definition. Adding another param is kinda not great as well.
And extending the api jyst for that use case is over the top imho :thinking:

**Comment by skjnldsv** [comment]
> 1. This is not clear for me when a file is in focus and when it is not. For example, for arrows and CTRL+Space hotkeys. Especially because the click on the file executes an action rather than selecting it.

Totally agree! But this is not really related to this PR.
The active file design is already a feature since 28 (the fileid in the URL initiate it)

File: apps/files/src/components/FilesListTableHeader.vue [hunk 2]
```diff
@@ -172,6 +181,32 @@ export default defineComponent({
...
+			if (isDialogOpened()) {
+				return
+			}
+
+			// ctrl+a selects all
+			if (event.key === 'a' && event.ctrlKey) {
+				this.onToggleAll(true)
+				event.preventDefault()
+				event.stopPropagation()
+			}
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply apps/files/src/components/FilesListTableHeader.vue]: Fixed with `useHotKey`

File: apps/files/src/views/Settings.vue [hunk 3]
```diff
@@ -145,6 +264,14 @@ export default {
 		},
 	},
 
+	created() {
+		// ? opens the settings dialog on the keyboard shortcuts section
+		useHotKey('?', this.showKeyboardShortcuts, {
+			stop: true,
+			prevent: true,
+		})
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply apps/files/src/views/Settings.vue:272]: Closing this, this is not related to that ticket. I'll open an issue on @nc/vue

File: apps/files/src/components/FilesListVirtual.vue [hunk 4]
```diff
@@ -243,21 +243,44 @@ export default defineComponent({
 		},
 	},
 
+	created() {
+		useHotKey('Escape', this.unselectFile, {
+			stop: true,
+			prevent: true,
+		})
+
+		useHotKey('ArrowUp', this.onKeyDown, {
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply apps/files/src/components/FilesListVirtual.vue:252]: Sorry, I'm reworking this, I should have adjusted the labels :)
I'll take that comment into consideration though, thanks! :+1:

**Review by artonge — APPROVED** [review]

**Comment by skjnldsv** [comment]
> Another option would be to have two actions, `delete` and `deleteWithConfirmation`

@artonge yeah, I might revisit this later tbh

**Comment by skjnldsv** [comment]
Failure unrelated :see_no_evil: 
image

**Review by nfebe — APPROVED** [review]
🚀🚀🚀

**Comment by skjnldsv** [comment]
image
 passes locally :thinking:

**Comment by marcoambrosini** [comment]
@skjnldsv I'm trying to test this on mac but it's not working. Is that supposed to be the cased?

**Pull request merged**